High-resolution app icon for MacOS

matteosecli opened the pull request:
Just a .icns version of https://github.com/TigerVNC/tigervnc/blob/master/media/tigervnc.svg

-----

Max resolution has been bumped from 512x512 to 1024x1024, and the underlying PNG quality has been increased. The icon is now consistent with all the other icons (it was slightly different, don't know why). Should now look much better on high-res screens:

<img width="1107" alt="Schermata 2020-07-04 alle 16 20 29" src="https://user-images.githubusercontent.com/4618521/86514469-b09d2980-be12-11ea-835c-558b94cba265.png">

CendioOssman commented:
Thank you. Unfortunately `media/tigervnc.svg` isn't really meant to be used as an icon. The one you are looking for is `media/icons/tigervnc.svg`.

Do you think you can generate a new file based on that?

matteosecli commented:
Yep, I've regenerated the icon based on `media/icons/tigervnc.svg`. This is the outcome (left: old icon | center: previously proposed hi-res icon | right: newly generated icon):

<img width="1604" alt="Schermata 2020-07-16 alle 16 54 52" src="https://user-images.githubusercontent.com/4618521/87687025-8722bd00-c785-11ea-8b92-efee5161e1e2.png">

There is some extra blank space on the border of this SVG file, hope this is the intended output.

Just as a future reference, since I've found ImageMagick's `convert` (both via the Inkscape and via the builtin backends) a bit unreliable on MacOS and producing poor results, this is the script I've used to generate the iconset, after having converted the `.svg` to a `png` via GIMP:

```bash
#!/usr/bin/env bash

# Convert tigervnc.svg to tigervnc.png, 1024x1024 px, 96 px/in, via GIMP.

mkdir tigervnc.iconset

cp tigervnc.png "tigervnc.iconset/icon_1024x1024.png"
cp "tigervnc.iconset/icon_1024x1024.png" "tigervnc.iconset/icon_512x512@2x.png"

convert -resize '!512x512' "tigervnc.png" "tigervnc.iconset/icon_512x512.png"
cp "tigervnc.iconset/icon_512x512.png" "tigervnc.iconset/icon_256x256@2x.png"

convert -resize '!256x256' "tigervnc.png" "tigervnc.iconset/icon_256x256.png"
cp "tigervnc.iconset/icon_256x256.png" "tigervnc.iconset/icon_128x128@2x.png"

convert -resize '!128x128' "tigervnc.png" "tigervnc.iconset/icon_128x128.png"
cp "tigervnc.iconset/icon_128x128.png" "tigervnc.iconset/icon_64x64@2x.png"

convert -resize '!64x64' "tigervnc.png" "tigervnc.iconset/icon_64x64.png"
cp "tigervnc.iconset/icon_64x64.png" "tigervnc.iconset/icon_32x32@2x.png"

convert -resize '!32x32' "tigervnc.png" "tigervnc.iconset/icon_32x32.png"
cp "tigervnc.iconset/icon_32x32.png" "tigervnc.iconset/icon_16x16@2x.png"

convert -resize '!16x16' "tigervnc.png" "tigervnc.iconset/icon_16x16.png"

iconutil -c icns tigervnc.iconset
```

I'll push the new icons in a few seconds.

CendioOssman commented:
> There is some extra blank space on the border of this SVG file, hope this is the intended output.

That's on purpose as we have a very square icon. Without the empty space it will look too large and out of place compared to other icons.



> Just as a future reference, since I've found ImageMagick's `convert` (both via the Inkscape and via the builtin backends) a bit unreliable on MacOS and producing poor results, this is the script I've used to generate the iconset, after having converted the `.svg` to a `png` via GIMP:

Nice, thanks. However it looks like you are using the same source SVG for every size? We have hand tweaked versions for 32x32 and 16x16. Do you think you could use those instead?

And shouldn't there be "normal DPI" versions included in the file as well?

matteosecli commented:
> Nice, thanks. However it looks like you are using the same source SVG for every size? We have hand tweaked versions for 32x32 and 16x16. Do you think you could use those instead?

Yes I could, but I would need the SVG source for those specific small-size icons (like the ones you already provide in `media/`), as I also have to provide `16x16` and `32x32` at double the pixel density as per Apple specifications, and with these hand-tweaked icons `16x16` at double the pixel density is visually different from `32x32` (and similarly `32x32@2x` is different from `64x64`).

> And shouldn't there be "normal DPI" versions included in the file as well?

What do you mean by "normal DPI"? I checked the DPI of the images in `media/` and seems to be 72 DPI; I could definitely upload an icon set generated with 72 DPI images just to be consistent with the rest (and I probably will, once I fix the `16x16` and `32x32` icons). But here it really makes no difference at all, because the number of pixels stays the same in both cases and the DPI value is just needed to convert px to cm or to in. Apple's developer instructions always specify the icon sizes in px and not in cm or in, so the DPI value is actually unused.

LicoMonch commented:
Some comments:
`# Convert tigervnc.svg to tigervnc.png, 1024x1024 px, 96 px/in, via GIMP.`
As you already have convert, why are you using gimp?
eg: convert -density 1200 -resize 200x200 source.svg target.png

I'd prefer native Apple tools to create files consumed by apple software:
```
mkdir tigervnc.iconset
for i in 16 32 128 256 512 1024
do
    [ ${i} -ne 1024 ] && /usr/bin/sips -z $i $i tigervnc.png --out "tigervnc.iconset/icon_${i}x${i}.png"
    /usr/bin/sips -z $i $i tigervnc.png --out "tigervnc.iconset/icon_$((i / 2))x$((i / 2))@2x.png"
done
/usr/bin/iconutil -c icns -o tigervnc.icns tigervnc.iconset
```

also see: https://developer.apple.com/library/archive/documentation/GraphicsAnimation/Conceptual/HighResolutionOSX/Optimizing/Optimizing.html
(it's a bit older but still relevant - afaik)

matteosecli commented:
> As you already have convert, why are you using gimp?

ImageMagick's `convert` uses Inkscape as a backend to convert from SVG to PNG. Before a Homebrew upgrade this produced poor results in terms of image quality, while a later Homebrew upgrade broke completely the svg to png conversion and I get a completely black image. There are multiple reports of this problem, it seems to be a combination of issues at least on MacOS. Not even upgrading to the latest Inkscape (and ImageMagick) solved the problem, no matter whichever combination of versions of the softwares I've tried.

If I bypass Inkscape and I force ImageMagick to use `librsvg` instead of Inkscape, the resulting image seems instead a Picasso version of the original SVG, completely broken.

I've also tried other (free) software and imho GIMP gives the best results with the SVG to PNG conversion, both in terms of image quality and filesize.

LicoMonch commented:
> > As you already have convert, why are you using gimp?
> 
> ImageMagick's `convert` uses Inkscape as a backend to convert from SVG to PNG. 

That confuses me even more ;)

Inkscape itself can do the trick without installing imagemagick or any other tool via brew ;)

```
curl -O https://raw.githubusercontent.com/TigerVNC/tigervnc/master/media/tigervnc.svg
/Applications/Inkscape.app//Contents/MacOS/inkscape --export-filename=tigervnc.png tigervnc.svg
```
reference: https://inkscape.org/doc/inkscape-man.html

(but that should not prevent anything here, gimp is okay too ;) )

LicoMonch commented:
Imho, anything here documented, could and maybe should be done at build-time of the .app
so you avoid having a blob (icns) in the source

matteosecli commented:
Yep, I'm indeed having issues with Inkscape, whether calling it via `convert` or directly... :)

matteosecli commented:
> Imho, anything here documented, could and maybe should be done at build-time of the .app
so you avoid having a blob (icns) in the source

I think they removed the build-time conversion from svg to png so not to have a build dependency on inkscape / imagemagick / whatever, which indeed makes sense. But yes, I agree that it would be nice to have it automated.

bphinz commented:
That would be preferable yes, but I too have had very inconsistent results
with convert on different platforms (including the all black results
mentioned) while producing the nightly & release builds

On Fri, Jul 17, 2020 at 7:35 AM LicoMonch <notifications@github.com> wrote:

> Imho, anything here documented, could and maybe should be done at
> build-time of the .app
> so you avoid having a blob (icns) in the source
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/TigerVNC/tigervnc/pull/1057#issuecomment-660058172>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB45M3JGOUJBGPCGSCPUZGTR4AZNVANCNFSM4OQNOFBA>
> .
>
-- 
Sent from Gmail Mobile

matteosecli commented:
@CendioOssman I've made the 16x16 and 32x32 SVG files myself and I've included them in the latest commit. I've used the corresponding `tigervnc_16.svg` and `tigervnc_28.svg` files in the `media/` folder and I've added the blank border so to match the other icons (if i.e. the icon is `48x48`, then the actual drawing is `42x42` + border). When converting these new SVG files to PNG, I see no differences between the existing PNG files in `media/icons/` and the images resulting from the conversion of the new SVGs (except for the fact that somehow the existing `tigervnc_16.png` is actually a `20x20` picture and doesn't have the extra border).

I've updated the `.icns` accordingly, and I've used 72 DPI for the actual images (as a side note on DPI: Inkscape used to work at 90 DPI, while the new version has switched to 96 DPI for better browser support. Anyway I've used 72 DPI here, which seems to be a "historical" value due to limitations of old screens)

CendioOssman commented:
The new icons look a lot better and has padding that fits in well with everything else. I compared with the system preferences app, which has similar icon shape as TigerVNC.

I'm not sure what the purpose of the new svg files are though. `tigervnc_32x32.svg` should be the same result as the existing `tigervnc_32.png`, and `tigervnc_16x16.svg` isn't that helpful as the point of having a 16x16 one was to make sure everything is pixel aligned. That isn't really a thing on macOS so it looks more like everything else if you just scale down something higher resolution. Or is there something else I'm overlooking?

matteosecli commented:
> I'm not sure what the purpose of the new svg files are though. `tigervnc_32x32.svg` should be the same result as the existing `tigervnc_32.png`,

Yep, that was indeed the goal. I needed to re-create the source SVG file because as per Apple specifications the icon file needs both a `32x32` icon (i.e. `tigervnc_32.png`) and the same `32x32` icon at double the resolution, i.e. `tigervnc_32@2x.png`, which unfortunately was not already provided (and just upscaling `tigervnc_32.png` would have resulted in a poor job). This icon is different from the `64x64` one because it has to match the appearance of the `32x32` one, so I could not use the already-provided `tigervnc.svg` for these smaller versions, because they differ a bit from their bigger parents.

> and `tigervnc_16x16.svg` isn't that helpful as the point of having a 16x16 one was to make sure everything is pixel aligned. That isn't really a thing on macOS so it looks more like everything else if you just scale down something higher resolution.

The story with the `16x16` icon was indeed a bit odd, because the original `tigervnc_16.png` is a `20x20` image and doesn't have the padding like the other images do. I re-created a `tigervnc_16x16.svg` with the addition of the extra padding so to match the other icons, but if you prefer so I can just ditch the `16x16` custom files and downscale the `32x32` icons so to use them as `16x16` icons as well. Just let me know what you prefer. 🙂

CendioOssman commented:
> Yep, that was indeed the goal. I needed to re-create the source SVG file because as per Apple specifications the icon file needs both a `32x32` icon (i.e. `tigervnc_32.png`) and the same `32x32` icon at double the resolution, i.e. `tigervnc_32@2x.png`, which unfortunately was not already provided (and just upscaling `tigervnc_32.png` would have resulted in a poor job). This icon is different from the `64x64` one because it has to match the appearance of the `32x32` one, so I could not use the already-provided `tigervnc.svg` for these smaller versions, because they differ a bit from their bigger parents.

Ah. But the source for `tigervnc_32.png` is `tigervnc_28.svg`. So you should be able to generate what you need from that.

> The story with the `16x16` icon was indeed a bit odd, because the original `tigervnc_16.png` is a `20x20` image and doesn't have the padding like the other images do.

What the... When did that happen? It's definitely not supposed to be like that.

> I re-created a `tigervnc_16x16.svg` with the addition of the extra padding so to match the other icons, but if you prefer so I can just ditch the `16x16` custom files and downscale the `32x32` icons so to use them as `16x16` icons as well. Just let me know what you prefer. slightly_smiling_face

In this case a downscale better matches how other applications look on macOS, so I think we should go with that.

CendioOssman commented:
@matteosecli, have you been able to have a look at those final details? Would be nice to get this merged. :)

matteosecli commented:
Hi @CendioOssman, been quite busy at work, lately; I'll give it a look within the next couple of days and push an updated version. 🙂

matteosecli commented:
> Ah. But the source for `tigervnc_32.png` is `tigervnc_28.svg`. So you should be able to generate what you need from that.

I've indeed used `tigervnc_28.svg` as a basis for `tigervnc_32x32.svg`, I've just added the extra blank padding so to match the look of the bigger-size `tigervnc.svg`. I guess one could also add an extra blank border via `convert` or even with `sips`, maybe by using something like `sips -p 32 32 --padColor FFFFFF` (not tested); if you prefer to go this route, feel free to delete the new 32x32 SVG source (not sure if you have write access, in that case I can delete the file). I just felt that having a plain SVG file that could be directly converted to the final PNG file without extra/multiple operations was a bit more clean. 🙂 

> In this case a downscale better matches how other applications look on macOS, so I think we should go with that.

Done, I've also deleted the corresponding 16x16 SVG file I created. 🙂

CendioOssman commented:
Thanks. Merged as a8229e17ebdbd649a789ca6fb283fdafa2f4997d.
